src/sensai/data_transformation.py

```diff
@@ -39,44 +100,73 @@ def applyInverse(self, df: pd.DataFrame) -> pd.DataFrame:
 class RuleBasedDataFrameTransformer(DataFrameTransformer, ABC):
     """Base class for transformers whose logic is entirely based on rules and does not need to be fitted to data"""
 
+    def _fit(self, df: pd.DataFrame):
+        pass
+
     def fit(self, df: pd.DataFrame):
         pass
 
+    def isFitted(self):
+        return True
 
-class DataFrameTransformerChain:
-    """Supports the application of a chain of data frame transformers"""
 
-    def __init__(self, dataFrameTransformers: Sequence[DataFrameTransformer]):
-        self.dataFrameTransformers = dataFrameTransformers
+class DataFrameTransformerChain(DataFrameTransformer):
+    """
+    Supports the application of a chain of data frame transformers.
+    During fit and apply each transformer in the chain receives the transformed output of its predecessor.
+    """
+    @staticmethod
+    def _flattened(l: Sequence[Union[DataFrameTransformer, Sequence[DataFrameTransformer]]]) -> List[DataFrameTransformer]:
+        result = []
+        for x in l:
+            if isinstance(x, DataFrameTransformer):
+                result.append(x)
+            else:
+                result.extend(x)
+        return result
 
-    def apply(self, df: pd.DataFrame, fit=False) -> pd.DataFrame:
-        """
-        Applies the chain of transformers to the given DataFrame, optionally fitting each transformer before applying it.
-        Each transformer in the chain receives the transformed output of its predecessor.
+    # TODO: do we really need the more flexible interface? I propose a breaking change to *dataFrameTransformers: DataFrameTransformer
+    #   without the flattened stuff
+    def __init__(self, *dataFrameTransformers: Union[DataFrameTransformer, List[DataFrameTransformer]]):
+        super().__init__()
+        self.dataFrameTransformers = self._flattened(dataFrameTransformers)
 
-        :param df: the data frame
-        :param fit: whether to fit the transformers before applying them
-        :return: the transformed data frame
-        """
+    def _apply(self, df: pd.DataFrame) -> pd.DataFrame:
         for transformer in self.dataFrameTransformers:
-            if fit:
-                transformer.fit(df)
             df = transformer.apply(df)
         return df
 
-    def fit(self, df: pd.DataFrame):
-        for transformer in self.dataFrameTransformers:
-            transformer.fit(df)
+    def _fit(self, df: pd.DataFrame):
```

Comment:
This changes the logic but I believe the previous logic was incorrect, right? Apply happens in a chain, so fit should do so as well

Reply:
Yes, definitely!